File: ClientLibs/Python/controller.py
import commutator
import commands
import protocol
class Controller:
    def __init__(self,host="localhost",port="2345",ID = 0):
        self.ID = str(ID)
        self.comm = commutator.Commutator(host,port)
        self.MoveCommand = commands.MoveCommand()
        self.RotateCommand = commands.RotateCommand()
        self.FireCommand = commands.FireCommand()
        self.ScanCommand = commands.ScanCommand()
    
    def state_process(self,state):
        if(state == -1):
            print "End of Game!"
            self.end_game()
            return -1
        if(state == 1):
                return 1
        if(state == 0):
                return 0
        return 0

    def ready(self):
        if(self.comm.up_connection()):
            return self.comm.hand_shake(protocol.Protocol.get_simple_msg_line(str(self.ID)))
        else:
            return False

    def end_game(self):
        return self.comm.down_connection(protocol.Protocol.get_simple_msg_line("EOG"))

    def move(self,destination):
        request = self.MoveCommand.produce_request(self.ID,destination)
        response = self.comm.exchange(request)
        return self.state_process(self.MoveCommand.parse_response(response))

    def rotate(self,destination):
        request = self.RotateCommand.produce_request(self.ID,destination)
        response = self.comm.exchange(request)
        return self.state_process(self.RotateCommand.parse_response(response))

    def fire(self):
        request = self.FireCommand.produce_request(self.ID)
        response = self.comm.exchange(request)
        return self.state_process(self.FireCommand.parse_response(response))

    def scan(self):
        request = self.ScanCommand.produce_request(self.ID)
        response = self.comm.exchange(request)
        answer = self.ScanCommand.parse_response(response)
        self.state_process(answer.state)
        return answer

File: ClientLibs/Haskell/Controller.hs
module Controller where
import Commands
import Commutator

type ID = String
data Controller = Controller ID Commutator

getController :: String -> String -> ID ->IO Controller
getController ip port id = do
						 comm <- getCommutator ip port
						 return $ Controller id comm

ready :: Controller -> IO Controller
ready ctrl = do
	comm' <- connectCommutator comm  
	handShake (id ++ "\r\n") comm'
	return $ Controller id comm'
	where Controller id comm  = ctrl

endGame :: Controller -> IO Controller
endGame ctrl = do
	comm' <- disconnectCommutator "EOG\r\n" comm
	return $ Controller id comm'
	where Controller id comm  = ctrl

--------------- Command functions ---------------  
produceCommand cmdName argument controller = do
						 let request = produceRequest id cmdName argument
						 response <- messageExchange request comm
						 return $ parseResponse response cmdName 
						 where 
						 	   Controller id comm = controller

move destination controller = produceCommand "MOV" (CommandArgument destination) controller
						 
rotate angle controller = produceCommand "ROT" (CommandArgument angle) controller

fire controller = produceCommand "FR" Nil controller

scan controller = produceCommand "SC" Nil controller
--------------- ---------------- ---------------
--------------- ---------------- ---------------
